src_lockfree/htbl/htbl.body.ml

```diff
@@ -0,0 +1,617 @@
+                }
+            in
+            if
+              Atomic_array.unsafe_compare_and_set r.buckets i (B before)
+                (B after)
+            then (
+              Size.update_once r.size cons_r.size_modifier;
+              cons_r.size_modifier <- Size.used_once;
+              adjust_size t r mask found_node)
+            else remove_node t key (Backoff.once backoff)
+      else
+        match dissoc r.equal key cons_r.rest with
+        | (Nil | Cons _) as rest ->
+            if
+              Atomic_array.unsafe_compare_and_set r.buckets i (B before)
+                (B (Cons_with_size { cons_r with rest; size_modifier }))
+            then (
+              Size.update_once r.size cons_r.size_modifier;
+              cons_r.size_modifier <- Size.used_once;
+              assoc_node r.equal key cons_r.rest |> adjust_size t r mask)
+            else remove_node t key (Backoff.once backoff)
+        | exception Not_found -> Nil
+    end
+  | B (Resize _) ->
+      let _ = finish t (Atomic.get t) in
+      remove_node t key Backoff.default
+
+let try_remove t key = remove_node t key Backoff.default != Nil
+
+let remove_exn t key =
+  match remove_node t key Backoff.default with
+  | Nil -> raise_notrace Not_found
+  | Cons r -> r.value
+
+(* *)
+
+let rec snapshot t ~clear backoff =
+  let r = get t in
+  if try_resize t r (Atomic_array.length r.buckets) ~clear then begin
+    (* At this point the resize has been completed and a new array is used for
+       buckets and [r.buckets] now has an immutable copy of what was in the hash
+       table. *)
+    let snapshot = r.buckets in
+    let rec loop i kvs () =
+      match kvs with
+      | Nil ->
+          if i = Atomic_array.length snapshot then Seq.Nil
+          else
+            loop (i + 1)
+              (match Atomic_array.unsafe_fenceless_get snapshot i with
+              | B (Resize { spine = Nil_with_size _ }) -> Nil
+              | B (Resize { spine = Cons_with_size cons_r }) ->
+                  Cons
+                    {
+                      key = cons_r.key;
+                      value = cons_r.value;
+                      rest = cons_r.rest;
+                    }
+              | B (Nil_with_size _ | Cons_with_size _) ->
+                  (* After resize only [Resize] values should be left in the old
+                     buckets. *)
+                  assert false)
+              ()
+      | Cons r -> Seq.Cons ((r.key, r.value), loop i r.rest)
+    in
+    loop 0 Nil
+  end
+  else snapshot t ~clear (Backoff.once backoff)
+
+let to_seq t = snapshot t ~clear:false Backoff.default
+let remove_all t = snapshot t ~clear:true Backoff.default
+
+(* *)
+
+let rec try_find_random_non_empty_bucket size buckets seed i =
+  match Atomic_array.unsafe_fenceless_get buckets i with
+  | B (Nil_with_size nil_r) | B (Resize { spine = Nil_with_size nil_r }) ->
+      if nil_r.size_modifier != Size.used_once then
+        Size.update_once size nil_r.size_modifier;
```

Comment:
There are cases where it probably improves performance not to do so, but usually you want to clear the `size_modifier` field by assigning `Size.used_once` to it:

```suggestion
      if nil_r.size_modifier != Size.used_once then begin
        Size.update_once size nil_r.size_modifier;
        nil_r.size_modifier <- Size.used_once
      end;
```